### ISSUE TITLE
[beam-refactor] StoreToZarr - Cannot Combine Fragments Possible Error

### ISSUE THREAD

**Issue by norlandrhagen** [pangeo-forge/pangeo-forge-recipes#517]
Hi there, I'm working through debugging a recipe using the beam-refactor branch. In the final step of the pipeline, `StoreToZarr` fails and gives an error: `ValueError: Cannot combine fragments. Expected a hypercube of shape [1] but got 4 fragments.`. 

Not sure if this is an actual error or some issue with my environment. I tried running the [multi_variable_recipe](https://pangeo-forge--516.org.readthedocs.build/en/516/pangeo_forge_recipes/tutorials/xarray_zarr/multi_variable_recipe.html) notebook, which also combines multiple variables into a single dataset and got the same error.

To create my environment, I ran:
```
mamba env create --file ci/py3.9.yml
mamba activate pangeo-forge-recipes
pip install git+https://github.com/pangeo-forge/pangeo-forge-recipes.git@beam-refactor
```

MRE:

```python

import os
from tempfile import TemporaryDirectory
import apache_beam as beam

from pangeo_forge_recipes.transforms import OpenURLWithFSSpec, OpenWithXarray, StoreToZarr, Indexed, T
from pangeo_forge_recipes.patterns import MergeDim, ConcatDim, FilePattern


start_year = 2000
end_year = 2001
variables = ["shum", "tmin", "tmax", "wind"]


def format_function(variable, time):
    return (f"https://rda.ucar.edu/thredds/dodsC/files/g/ds314.0/0.25deg/daily/{variable}_0p25_daily_{time}-{time}.nc")

class DropZDim(beam.PTransform):
    """A preprocessing function to drop z dim"""

    @staticmethod
    def _drop_z_dim(item: Indexed[T]) -> Indexed[T]:
        index, ds = item
        ds = ds.squeeze("z").drop('z')
        return index, ds

    def expand(self, pcoll: beam.PCollection) -> beam.PCollection:
        return pcoll | beam.Map(self._drop_z_dim)


concat_dim = ConcatDim("time",keys=range(start_year, end_year + 1),nitems_per_file=1)
variable_merge_dim = MergeDim("variable", keys=variables)
pattern = FilePattern(format_function, variable_merge_dim, concat_dim, file_type='opendap')


td = TemporaryDirectory()
target_root = td.name
store_name = "output.zarr"
target_store = os.path.join(target_root, store_name)

transforms = (
    beam.Create(pattern.items())
    | OpenWithXarray(file_type=pattern.file_type)
    | DropZDim()
    | StoreToZarr(
        store_name=store_name,
        target_root=target_root,
        combine_dims=pattern.combine_dim_keys,
    )
)

with beam.Pipeline() as p:
    p | transforms

```

Traceback:

<details>

```python 
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File [~/opt/anaconda3/envs/install/envs/pangeo-forge-recipes/lib/python3.9/site-packages/apache_beam/runners/common.py:1418](https://file+.vscode-resource.vscode-cdn.net/Users/nrhagen/Documents/carbonplan/pangeo_forge/pangeo-forge-recipes/docs/pangeo_forge_recipes/tutorials/~/opt/anaconda3/envs/install/envs/pangeo-forge-recipes/lib/python3.9/site-packages/apache_beam/runners/common.py:1418), in apache_beam.runners.common.DoFnRunner.process()

File [~/opt/anaconda3/envs/install/envs/pangeo-forge-recipes/lib/python3.9/site-packages/apache_beam/runners/common.py:625](https://file+.vscode-resource.vscode-cdn.net/Users/nrhagen/Documents/carbonplan/pangeo_forge/pangeo-forge-recipes/docs/pangeo_forge_recipes/tutorials/~/opt/anaconda3/envs/install/envs/pangeo-forge-recipes/lib/python3.9/site-packages/apache_beam/runners/common.py:625), in apache_beam.runners.common.SimpleInvoker.invoke_process()

File [~/opt/anaconda3/envs/install/envs/pangeo-forge-recipes/lib/python3.9/site-packages/apache_beam/transforms/core.py:1977](https://file+.vscode-resource.vscode-cdn.net/Users/nrhagen/Documents/carbonplan/pangeo_forge/pangeo-forge-recipes/docs/pangeo_forge_recipes/tutorials/~/opt/anaconda3/envs/install/envs/pangeo-forge-recipes/lib/python3.9/site-packages/apache_beam/transforms/core.py:1977), in MapTuple..(x)
   1976 else:
-> 1977   wrapper = lambda x: [fn(*x)]
   1979 # Proxy the type-hint information from the original function to this new
   1980 # wrapped function.

File [~/opt/anaconda3/envs/install/envs/pangeo-forge-recipes/lib/python3.9/site-packages/pangeo_forge_recipes/rechunking.py:185](https://file+.vscode-resource.vscode-cdn.net/Users/nrhagen/Documents/carbonplan/pangeo_forge/pangeo-forge-recipes/docs/pangeo_forge_recipes/tutorials/~/opt/anaconda3/envs/install/envs/pangeo-forge-recipes/lib/python3.9/site-packages/pangeo_forge_recipes/rechunking.py:185), in combine_fragments(group, fragments)
    183 if len(fragments) != total_size:
    184     # this error path is currently untested
--> 185     raise ValueError(
    186         "Cannot combine fragments. "
    187         f"Expected a hypercube of shape {shape} but got {len(fragments)} fragments."
    188     )
    190 starts_cube = [np.array(item[1]).reshape(shape) for item in dims_starts_sizes]

ValueError: Cannot combine fragments. Expected a hypercube of shape [1] but got 4 fragments.

During handling of the above exception, another exception occurred:
...
    188     )
    190 starts_cube = [np.array(item[1]).reshape(shape) for item in dims_starts_sizes]
    191 sizes_cube = [np.array(item[2]).reshape(shape) for item in dims_starts_sizes]

ValueError: Cannot combine fragments. Expected a hypercube of shape [1] but got 4 fragments. [while running '[9]: Create|OpenWithXarray|DropZDim|StoreToZarr[/StoreToZarr/Rechunk/MapTuple](https://file+.vscode-resource.vscode-cdn.net/StoreToZarr/Rechunk/MapTuple)(combine_fragments)']
```
</details>

**Comment by cisaacstern** [comment]
Possibly related to https://github.com/pangeo-forge/pangeo-forge-recipes/issues/504?

**Comment by cisaacstern** [comment]
@norlandrhagen any chance passing decode_coords="all" to OpenWithXarray, as suggested in https://github.com/pangeo-forge/pangeo-forge-recipes/issues/504#issuecomment-1485235403, solves this issue?

**Comment by norlandrhagen** [comment]
Just tried adding `xarray_open_kwargs={"decode_coords": "all"}` to `OpenWithXarray` and got the same error as above. 😕

**Comment by norlandrhagen** [comment]
For future ref, the dataset repr:
<img width="857" alt="image" src="https://user-images.githubusercontent.com/22455466/234683853-79659fab-1782-4107-99fe-306f4d4f0620.png">

**Comment by rabernat** [comment]
>  Expected a hypercube of shape [1] but got 4 fragments.

I'm wondering if it's  more than a coincidence that there are 4 variables in the MergeDim. 🤔

**Comment by norlandrhagen** [comment]
Yup definitely corresponds to the # of variables! The recipe runs fine with a single variable and fails with two variables with: `Expected a hypercube of shape [1] but got 2 fragments`.

**Comment by norlandrhagen** [comment]
One possible clue is that this [docs build](https://pangeo-forge--500.org.readthedocs.build/en/500/pangeo_forge_recipes/tutorials/xarray_zarr/multi_variable_recipe.html)  from PR #500 for the [`multi_variable_recipe`](https://pangeo-forge--500.org.readthedocs.build/en/500/pangeo_forge_recipes/tutorials/xarray_zarr/multi_variable_recipe.html) succeeded, while running that recipe locally fails with the same `Expected a hypercube of shape [1] but got 2 fragments` error.

**Comment by rabernat** [comment]
Ok so that suggests where you can look to debug and fix this. The signature of `combine_fragements`:

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/f10ec3cb3c8272a40c06731afee0eed5a93f468e/pangeo_forge_recipes/rechunking.py#L132-L142

says explicitly:

> Only combines concat dims; merge dims are not combined

I would start by trying to write a test in [test_rechunking.py](https://github.com/pangeo-forge/pangeo-forge-recipes/blob/beam-refactor/tests/test_rechunking.py) which reproduces the same error.

You'll notice that the existing tests don't include any merge dims (except for an error case).

**Comment by norlandrhagen** [comment]
Ah nice! I'll try to add a test. Thanks for the feedback.

**Comment by rabernat** [comment]
Awesome! Definitely ping if you get stuck. This will be a good exercise which will force you to grok the important concepts of `Dimension`, `CombineOp.CONCAT`, and `Index`. Making up synthetic data for a test is an excellent way to understand how the internals work.

**Comment by rabernat** [comment]
Ah, one more thought. Perhaps it's a bug that these four different fragments are even getting to `combine_fragments`. Look at how the `Rechunk` transform works

https://github.com/pangeo-forge/pangeo-forge-recipes/blob/f10ec3cb3c8272a40c06731afee0eed5a93f468e/pangeo_forge_recipes/transforms.py#L257-L267

Basically, it splits up the original dataset pieces into fragments, and then puts them back together. Each merge dim should end up in a separate group. That suggests you should also be looking at `split_fragment`.

**Comment by rabernat** [comment]
Was this fixed by #521?

**Comment by cisaacstern** [comment]
> Was this fixed by #521?

Yes, closing now 😀
